sdks/go/pkg/beam/core/runtime/exec/pcollection.go

```diff
@@ -93,9 +99,19 @@ func (p *PCollection) ProcessElement(ctx context.Context, elm *FullValue, values
 		} else {
 			p.nextSampleIdx = cur + p.r.Int63n(cur/10+2) + 1
 		}
-		var w byteCounter
-		p.elementCoder.Encode(elm, &w)
-		p.addSize(int64(w.count))
+
+		if p.dataSampler == nil {
+			var w byteCounter
+			p.elementCoder.Encode(elm, &w)
+			p.addSize(int64(w.count))
+		} else {
+			var buf bytes.Buffer
+			EncodeWindowedValueHeader(p.windowCoder, elm.Windows, elm.Timestamp, elm.Pane, &buf)
+			winSize := buf.Len()
+			p.elementCoder.Encode(elm, &buf)
```

Comment:
For current implementation in pcollection.go, Pcollection.Up() will panic if WindowCoder is nil. Based on the code in makeCoderForPCollection, it only returns nil windowCoder with an error which should be caught before init pCollection node.

Reply:
By definition, all PCollections in Beam are WindowValue encoded, so it's an error for there not to be a windowed value coder. The panic exists for debugging mistakes like that early, rather than later in processing.